R/eval-tidy.R

```diff
@@ -342,3 +339,16 @@ f_self_eval <- function(overscope, overscope_top) {
     .Call(rlang_eval, f_rhs(f), overscope)
   }
 }
+tilde_eval <- function(f, env) {
+  if (is_env(attr(f, ".Environment"))) {
+    return(f)
+  }
+  f <- duplicate(f, shallow = TRUE)
+
+  # Inline the base primitive because overscopes override `~` to make
+  # quosures self-evaluate
+  mut_node_car(f, base::`~`)
```

Comment:
that's what we are now doing, constructing a new call (by shallow duplication). I don't think we need full duplication. On the other hand that's only for bare formulas, not for quosures, so it doesn't matter as much. Also we need to inline the primitive function because `~` has a different definition in the overscope.